eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala

```diff
@@ -1126,81 +1184,65 @@ object Helpers {
       // is our main output confirmed (if we have one)?
-      val isMainOutputConfirmed = localCommitPublished.claimMainDelayedOutputTx.forall(tx => confirmedTxs.contains(tx.txid))
-      // are all htlc outputs from the commitment tx spent?
-      val unspentCommitTxHtlcOutputs = commitments.localCommit.publishableTxs.htlcTxsAndSigs.map(_.txinfo.input.outPoint).toSet -- localCommitPublished.irrevocablySpent.keys
+      val isMainOutputConfirmed = localCommitPublished.claimMainDelayedOutputTx.forall(tx => localCommitPublished.irrevocablySpent.contains(tx.input.outPoint))
+      // are all htlc outputs from the commitment tx spent (we need to check them all because we may receive preimages later)?
+      val allHtlcsSpent = (localCommitPublished.htlcTxs.keySet -- localCommitPublished.irrevocablySpent.keys).isEmpty
       // are all outputs from htlc txs spent?
-      val unconfirmedHtlcDelayedTxs = localCommitPublished.claimHtlcDelayedTxs
+      val unconfirmedHtlcDelayedTxs = localCommitPublished.claimHtlcDelayedTxs.map(_.input.outPoint)
         // only the txs which parents are already confirmed may get confirmed (note that this eliminates outputs that have been double-spent by a competing tx)
-        .filter(tx => (tx.txIn.map(_.outPoint.txid).toSet -- confirmedTxs).isEmpty)
+        .filter(input => confirmedTxs.contains(input.txid))
         // has the tx already been confirmed?
-        .filterNot(tx => confirmedTxs.contains(tx.txid))
-      isCommitTxConfirmed && isMainOutputConfirmed && unspentCommitTxHtlcOutputs.isEmpty && unconfirmedHtlcDelayedTxs.isEmpty
+        .filterNot(input => localCommitPublished.irrevocablySpent.contains(input))
+      isCommitTxConfirmed && isMainOutputConfirmed && allHtlcsSpent && unconfirmedHtlcDelayedTxs.isEmpty
     }
 
     /**
      * A remote commit is considered done when all commitment tx outputs that we can spend have been spent and confirmed
      * (even if the spending tx was not ours).
      */
-    def isRemoteCommitDone(keyManager: ChannelKeyManager, remoteCommitPublished: RemoteCommitPublished, commitments: Commitments): Boolean = {
-      val remoteCommit = commitments.remoteNextCommitInfo match {
-        case Left(WaitingForRevocation(nextRemoteCommit, _, _, _)) if nextRemoteCommit.txid == remoteCommitPublished.commitTx.txid => nextRemoteCommit
-        case _ => commitments.remoteCommit
-      }
-      val (_, htlcTimeoutTxs, htlcSuccessTxs) = Commitments.makeRemoteTxs(
-        keyManager,
-        commitments.channelVersion,
-        remoteCommit.index,
-        commitments.localParams,
-        commitments.remoteParams,
-        commitments.commitInput,
-        remoteCommit.remotePerCommitmentPoint,
-        remoteCommit.spec
-      )
-      val confirmedTxs = remoteCommitPublished.irrevocablySpent.values.toSet
+    def isRemoteCommitDone(remoteCommitPublished: RemoteCommitPublished): Boolean = {
```

Comment:
Yeah, I'm glad I could quickly remove these ugly additional parameters :sweat_smile: 

This could indeed move inside the `CommitPublished` types: that's what we've done in `eclair-kmp`, where we've put a lot more logic in these types. I'll explore this.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1728/commits/a4fbc3aaeda5037bcf3832fcb82f65ed869395de